### ISSUE TITLE
`The Pulumi CLI does not support transforms` exception in tests using `Mocks`

### ISSUE THREAD

**Issue by tlinhart** [pulumi/pulumi#17090]
### What happened?

I'm trying to migrate from `pulumi.runtime.register_stack_transformation` to `pulumi.runtime.register_resource_transform` in the [pulumi-aws-tags](https://github.com/tlinhart/pulumi-aws-tags) library. However, it throws the following exception in the `pytest` tests:

```
ERROR python/tests/test_autotag.py - Exception: The Pulumi CLI does not support transforms. Please update the Pulumi CLI.
```

### Example

Here's a standalone example that leads to the above error:

```python
import pulumi
import pulumi_aws as aws


class PulumiMocks(pulumi.runtime.Mocks):
    def new_resource(self, args: pulumi.runtime.MockResourceArgs):
        outputs = {**args.inputs}
        return [f"{args.name}_id", outputs]

    def call(self, args: pulumi.runtime.MockCallArgs):
        return {}


pulumi.runtime.set_mocks(PulumiMocks())


def _tags(args, tags):
    args.props["tags"] = {**(args.props.get("tags") or {}), **tags}
    return pulumi.ResourceTransformResult(args.props, args.opts)


pulumi.runtime.register_resource_transform(
    lambda args: _tags(args, {"foo": "bar"})
)

server = aws.ec2.Instance(
    "server",
    instance_type="t2.micro",
    ami="ami-0bb75d95f668ff5a7",
    tags={"group": "webservers"},
)


@pulumi.runtime.test
def test_server_has_tags():
    def check_tags(args):
        urn, tags = args
        assert tags is not None, f"server {urn} must have tags"
        assert "foo" in tags, f"server {urn} must have 'foo' tag"
        assert "group" in tags, f"server {urn} must have 'group' tag"

    return pulumi.Output.all(server.urn, server.tags).apply(check_tags)
```

### Output of `pulumi about`

```
CLI          
Version      3.130.0
Go Version   go1.22.6
Go Compiler  gc

Host     
OS       ubuntu
Version  24.04
Arch     x86_64

Backend        
Name           xxx
URL            file:///home/xxx
User           xxx
Organizations  
Token type     personal

Pulumi locates its logs in /tmp by default
```

Pulumi Python packages:

```
pulumi==3.130.0
pulumi_aws==6.50.1
```

### Additional context

It started with [this issue](https://github.com/tlinhart/pulumi-aws-tags/issues/23) which led me to [this blog post](https://www.pulumi.com/blog/resource-transforms/) explicitely noting the `pulumi-eks` package not being supported by stack transformations but supported by new transforms.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
The new transform system is implemented in the engine, and so there isn't currently a good way of running them in mocked tests currently. The most obvious way would be to reimplement the transform system in each language so it can run as part of mocks. That's not seen as very sustainable way of implementing it, so we've been thinking of if there's a simpler way to support these functions in tests.

**Comment by tlinhart** [comment]
Thanks for answer. What would you suggest as a best approach now? Is there any other issue or another place where I could track this problem so I can release a new version when there are ways to cover this in tests?

**Comment by Frassle** [comment]
> What would you suggest as a best approach now? 

There probably isn't a "best" fix. Its either use the old system with support for tests, or the new system with support for eks.

> Is there any other issue or another place where I could track this problem so I can release a new version when there are ways to cover this in tests?

I don't think there's another issue, we can use this ticket to track it.

**Comment by darthkurak** [comment]
Isn't this a breaking change? I have BaseStack for all of my stacks, which has some basic Transformations (like auto-tagging). I have Unit Tests for it, which play a critical role here. Right now, after migrating to Transforms, I cannot test these basic mechanisms. Any tip on how to workaround it?
Write integration tests? But API for integration tests does not return the created resources list - I would have to fetch them directly from the cloud provider somehow... which is much more painful to do than working with a created resources list.

**Comment by tlinhart** [comment]
@darthkurak moving from unit tests to integration testing (with LocalStack and Automation API) is the path I'm taking, see [here](https://github.com/tlinhart/pulumi-aws-tags/commit/a208122e855c4a4bd8a9b4e1c9a49494a1748acb).

**Comment by Frassle** [comment]
> Isn't this a breaking change?

No if the old system is working for you keep using it. We're not planning on removing it and its still supported, there's just somethings it will never support (like awsx and similar component libraries). If you _need_ awsx support then you need to use the new system, but that's going to be new code and currently it comes with the downside that it can't be unit tested.

**Comment by darthkurak** [comment]
> > Isn't this a breaking change?
> 
> No if the old system is working for you keep using it. We're not planning on removing it and its still supported, there's just somethings it will never support (like awsx and similar component libraries). If you _need_ awsx support then you need to use the new system, but that's going to be new code and currently it comes with the downside that it can't be unit tested.

But you announced that old API can be deprecated in the future. Does this mean that before you do, you will ensure that the new system will be testable in UnitTests?

**Comment by Frassle** [comment]
> But you announced that old API can be deprecated in the future. Does this mean that before you do, you will ensure that the new system will be testable in UnitTests?

Yes. We do intend to deprecate the old API but _in the future_ when the new API can cover everything.

**Comment by Frassle** [comment]
Currently wondering if it would be enough for the mock system to report that it _does_ support transforms (so that register transforms doesn't throw), and possibly exposing a register_transform method on the Mock interface so you can test transforms are registered and call them in your implementation of new_resource if wanted.

The first part is very easy, just need to fix some bool flags in the mock setup. 

The second part needs some grpc wrapping to handle the remote nature of transforms, but probably isn't too much work to do for each language.

We have no plans to make the mock system call registered transforms automatically. The whole point of the mock system is it's testing the boundary between your program and the engine. The user is supposed to mock out the engine response they expect, we're not going to start moving engine logic into the mock system to do some of that automatically.

**Comment by tlinhart** [comment]
For my use case it's needed that the transform actually runs. But I understand your point and found a way to overcome this (using LocalStack and actually provisioning the resources via automation API).

**Comment by tgummerer** [comment]
Gonna close this out, as it's done with https://github.com/pulumi/pulumi/pull/24416, https://github.com/pulumi/pulumi/pull/24406, https://github.com/pulumi/pulumi/pull/24415 and https://github.com/pulumi/pulumi-dotnet/pull/1139.  Docs are in a draft PR https://github.com/pulumi/docs/pull/21081, which is gonna be merged once the above PRs all make it into a release.
